File: ClearAzureQueues/Models/QueueSelectionModel.cs
using Microsoft.WindowsAzure.Storage.Queue;
using System;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Windows;
using System.Windows.Data;

namespace ClearAzureQueues.Models {

    public class QueueSelectionSettings {
        public AccountSettings Account { get; set; }
        public string NameFilter { get; set; }
    }

    public class QueueSelectionModel : DependencyObject {

        public QueueSelectionSettings Settings {
            get {
                return new QueueSelectionSettings() {
                    Account = Account.Settings,
                    NameFilter = NameFilter
                };
            }
        }

        public static readonly DependencyProperty AccountProperty =
            DependencyProperty.Register("Account", typeof(AccountModel), typeof(QueueSelectionModel));
        public AccountModel Account {
            get { return (AccountModel)GetValue(AccountProperty); }
            set { SetValue(AccountProperty, value); }
        }

        public static readonly DependencyProperty QueuesProperty =
            DependencyProperty.Register("Queues", typeof(ObservableCollection<QueueModel>), typeof(QueueSelectionModel));
        public ObservableCollection<QueueModel> Queues {
            get { return (ObservableCollection<QueueModel>)GetValue(QueuesProperty); }
            set { SetValue(QueuesProperty, value); }
        }

        public static readonly DependencyProperty FilteredQueuesProperty =
            DependencyProperty.Register("FilteredQueues", typeof(ListCollectionView), typeof(QueueSelectionModel));
        public ListCollectionView FilteredQueues {
            get { return (ListCollectionView)GetValue(FilteredQueuesProperty); }
            set { SetValue(FilteredQueuesProperty, value); }
        }

        public static readonly DependencyProperty NameFilterProperty =
            DependencyProperty.Register("NameFilter", typeof(string), typeof(QueueSelectionModel),
                new FrameworkPropertyMetadata((sender,e) => {
                    var model = sender as QueueSelectionModel;
                    if (model != null) model.Filter();
                }));
        public string NameFilter {
            get { return (string)GetValue(NameFilterProperty); }
            set { SetValue(NameFilterProperty, value); }
        }

        public static readonly DependencyProperty SelectedQueueProperty =
            DependencyProperty.Register("SelectedQueue", typeof(QueueModel), typeof(QueueSelectionModel));
        public QueueModel SelectedQueue {
            get { return (QueueModel)GetValue(SelectedQueueProperty); }
            set { SetValue(SelectedQueueProperty, value); }
        }

        public QueueSelectionModel(AccountModel account) {
            Queues = new ObservableCollection<QueueModel>();
            FilteredQueues = new ListCollectionView(Queues);
            Account = account;
            Populate();
        }

        public QueueSelectionModel(QueueSelectionSettings settings) : this(new AccountModel(settings.Account)) {
            NameFilter = settings.NameFilter;
        }

        public void Populate() {
            if (Account.IsConnected) {
                var worker = new BackgroundWorker();
                worker.WorkerReportsProgress = true;
                worker.DoWork += (sender, e) => {
                    var client = (CloudQueueClient)e.Argument;
                    foreach (var queue in client.ListQueues()) {
                        worker.ReportProgress(0, queue);
                    }
                };
                worker.ProgressChanged += (sender, e) => {
                    var queue = (CloudQueue)e.UserState;
                    foreach (var existing in Queues)
                        if (existing.QueueName == queue.Name)
                            return;
                    Queues.Add(new QueueModel(queue));
                };
                worker.RunWorkerCompleted += (sender, e) => {
                };
                worker.RunWorkerAsync(Account.Client);
            }
        }

        private void Filter() {
            if (String.IsNullOrEmpty(NameFilter)) {
                FilteredQueues.Filter = null;
            } else {
                FilteredQueues.Filter = new Predicate<object>(x => {
                    var model = x as QueueModel;
                    return
                        model != null &&
                        model.QueueName != null &&
                        NameFilter.Split(' ').All(model.QueueName.Contains);
                });
            }
        }
    }
}

File: ClearAzureQueues/Models/QueueSelectionModel.fs
namespace ClearAzureQueues.Models

open System
open System.Collections.ObjectModel
open System.ComponentModel
open System.Linq
open System.Windows
open System.Windows.Data
open Microsoft.WindowsAzure.Storage.Queue

open ClearAzureQueues
open ClearAzureQueues.Settings

[<AllowNullLiteral>]
type QueueSelectionModel(account:AccountModel) =
    inherit DependencyObject()

    let queues = new ObservableCollection<QueueModel>()
    let filteredQueues = new ListCollectionView(queues)

    let populate() =
        match account.Client with
        | Some client when account.IsConnected ->
            let worker = new BackgroundWorker()
            worker.WorkerReportsProgress <- true
            worker.DoWork.Add(fun e ->
                for queue in client.ListQueues() do
                    worker.ReportProgress(0, queue))
            worker.ProgressChanged.Add(fun e ->
                let queue = e.UserState :?> CloudQueue
                let exists =
                    queues
                    |> Seq.exists (fun x -> x.QueueName = queue.Name)
                if not exists then
                    let queueModel = new QueueModel(queue)
                    queueModel.UpdateStatus()
                    queues.Add(queueModel))
            worker.RunWorkerAsync()
        | _ -> ()

    do populate()

    static let selectedQueue =
        DependencyProperty.Register("SelectedQueue", typeof<QueueModel>, typeof<QueueSelectionModel>)
    static let nameFilter =
        DependencyProperty.Register("NameFilter", typeof<string>, typeof<QueueSelectionModel>,
            new FrameworkPropertyMetadata(new PropertyChangedCallback(fun (sender:DependencyObject) _ ->
                match sender with
                | :? QueueSelectionModel as model -> model.Filter()
                | _ -> ())))

    new { Account = account
          NameFilter = nameFilter } as x =
        QueueSelectionModel(new AccountModel(account)) then
            x.NameFilter <- nameFilter

    member public x.Settings =
        { Account = account.Settings
          NameFilter = x.NameFilter }

    member public x.NameFilter
        with get() = x.GetValue(nameFilter) :?> string
        and set(value:string) = x.SetValue(nameFilter, value)

    member public x.SelectedQueue
        with get() = match x.GetValue(selectedQueue) with :? QueueModel as m -> m | _ -> null
        and set(value:QueueModel) = x.SetValue(selectedQueue, value)

    member public x.Account = account

    member public x.Queues = queues

    member public x.FilteredQueues = filteredQueues

    member public x.Populate() = populate()

    member public x.Filter() =
        if String.IsNullOrWhiteSpace(x.NameFilter) then
            x.FilteredQueues.Filter <- null
        else
            let matches : string->bool =
                let terms = x.NameFilter.Split(' ')
                function
                | x when String.IsNullOrWhiteSpace(x) -> false
                | x -> Seq.forall x.Contains terms
            x.FilteredQueues.Filter <-
                new Predicate<obj>(function
                    | :? QueueModel as model -> matches model.QueueName
                    | _ -> false)